### ISSUE TITLE
Add method to get relative paths (for paths)

### ISSUE THREAD

**Issue by patriksvensson** [cake-build/cake#621]
It would be nice (especially for logs) if `DirectoryPath` and `FilePath` had a `GetRelativePath` method that returned a new path that is relative to another path.

```csharp
public class FilePath {
  public FilePath GetRelativePath(FilePath to) {
    // Use file's directory as from path.
  }
  public DirectoryPath GetRelativePath(DirectoryPath to) {
    // Use file's directory as from path.
  }
}
```

```csharp
public class DirectoryPath {
  public FilePath GetRelativePath(FilePath to) {
    ...
  }
  public DirectoryPath GetRelativePath(DirectoryPath to) {
    ...
  }
}
```

**Requirements**

- Full unit test coverage proving that the functionality works as expected.
- No PInvoke calls to Shlwapi.dll to get the relative path.
- No URI tricks to get the relative path.
- Both from/to paths must be absolute.
- If no relative path can be calculated, the method should throw.

**Basic test cases**

```csharp
var path = new DirectoryPath("C:/A/D/E");
var result = path.GetRelativePath(new DirectoryPath("C:/A/B/C"));
Assert.Equal("../../B/C", result.FullPath);
```

```csharp
var path = new DirectoryPath("D:/A/B/C");
var result = Record.Exception(() => path.GetRelativePath(new DirectoryPath("C:/D/E/F")));
Assert.IsInstanceOf<InvalidOperationException>(result);
Assert.Equal("Paths must share a common prefix.", result.Message);
```

```csharp
var path = new DirectoryPath("C:/A/B/C/D/E/F");
var result = path.GetRelativePath(new DirectoryPath("C:/A/B/C"));
Assert.Equal("../../..", result.FullPath);
```

```csharp
var path = new DirectoryPath("C:/A/B/C");
var result = path.GetRelativePath(new DirectoryPath("C:/A/B/C/D/E/F"));
Assert.Equal("./D/E/F", result.FullPath);
```

**Comment by ErikSchierboom** [comment]
@patriksvensson I am looking into this feature. What do you expect the returned FilePath to be for this code:

```csharp
var path = new FilePath("C:/hello.txt");
var result = path.GetRelativePath(new FilePath("C:/hello.txt"));
result = ???
```

Should it be `"."`, or `"./hello.txt"`, or just `"hello.txt"`?

**Comment by ErikSchierboom** [comment]
And shouldn't the `GetRelativePath(DirectoryPath to)` method return a `DirectoryPath` instead of a `Directory`?

**Comment by patriksvensson** [comment]
@ErikSchierboom 

Great that you're looking at this! :+1: 
Not 100% sure about the specification I wrote though, since it was only something i threw together when I wanted the functionality, but I **think** this is what we want though. What do you think? Something you see that would improve it?

1. Do you mean the `FullPath` property of the FilePath/DirectoryPath that is returned? Creating a path with or without `./` will result in the same thing (a relative file/directory path without the `./` part).

2. My bad. It should indeed return a `DirectoryPath`. I've updated it. :smile:

**Comment by daveaglick** [comment]
@ErikSchierboom Funny - I was literally *just* thinking about this. I'm porting `Path`, `FilePath`, and `DirectoryPath` into an unrelated project that needs to calculate relative paths. If it helps, I have some code that does something similar based on `System.IO.Path` here: https://github.com/Wyamio/Wyam/blob/develop/src/Wyam.Common/IO/PathHelper.cs#L32

**Comment by ErikSchierboom** [comment]
@patriksvensson That answers my questions, thanks!

@daveaglick I'll have a look! Thanks

**Comment by patriksvensson** [comment]
Resolved via #650.
